src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/ritual/RitualManager.java

```diff
@@ -0,0 +1,99 @@
+import com.github.minecraftschurlimods.arsmagicalegacy.common.ritual.trigger.GameEventRitualTrigger;
+import com.github.minecraftschurlimods.arsmagicalegacy.common.ritual.trigger.ItemDropRitualTrigger;
+import com.github.minecraftschurlimods.codeclib.CodecDataManager;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.mojang.serialization.Codec;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.common.util.Lazy;
+import org.apache.logging.log4j.LogManager;
+
+/**
+ *
+ */
+public final class RitualManager extends CodecDataManager<Ritual> {
+    private static final Lazy<RitualManager> INSTANCE = Lazy.concurrentOf(RitualManager::new);
+
+    private static final BiMap<ResourceLocation, Codec<? extends RitualTrigger>>     ritualTriggerCodecs     = HashBiMap.create();
+    private static final BiMap<ResourceLocation, Codec<? extends RitualRequirement>> ritualRequirementCodecs = HashBiMap.create();
+    private static final BiMap<ResourceLocation, Codec<? extends RitualEffect>>      ritualEffectCodecs      = HashBiMap.create();
+
+    private RitualManager() {
+        super("am_rituals", Ritual.CODEC, LogManager.getLogger());
+        useRegistryOps();
+    }
+
+    public static RitualManager instance() {
+        return INSTANCE.get();
+    }
+
+    public static void registerRitualRequirement(ResourceLocation id, Codec<? extends RitualRequirement> codec) {
+        ritualRequirementCodecs.put(id, codec);
+    }
+
+    public static void registerRitualEffect(ResourceLocation id, Codec<? extends RitualEffect> codec) {
+        ritualEffectCodecs.put(id, codec);
+    }
+
+    public static void registerRitualTrigger(ResourceLocation id, Codec<? extends RitualTrigger> codec) {
+        ritualTriggerCodecs.put(id, codec);
+    }
+
+    public static ResourceLocation getRitualRequirementType(RitualRequirement ritualRequirement) {
+        return ritualRequirementCodecs.inverse().get(ritualRequirement.codec());
+    }
+
+    public static Codec<? extends RitualRequirement> getRitualRequirementCodec(ResourceLocation resourceLocation) {
+        return ritualRequirementCodecs.get(resourceLocation);
+    }
+
+    public static ResourceLocation getRitualTriggerType(RitualTrigger ritualTrigger) {
+        return ritualTriggerCodecs.inverse().get(ritualTrigger.codec());
+    }
+
+    public static Codec<? extends RitualTrigger> getRitualTriggerCodec(ResourceLocation resourceLocation) {
+        return ritualTriggerCodecs.get(resourceLocation);
+    }
+
+    public static ResourceLocation getRitualEffectType(RitualEffect ritualEffect) {
+        return ritualEffectCodecs.inverse().get(ritualEffect.codec());
+    }
+
+    public static Codec<? extends RitualEffect> getRitualEffectCodec(ResourceLocation resourceLocation) {
+        return ritualEffectCodecs.get(resourceLocation);
+    }
+
+    static {
```

Comment:
I really dislike this static block. Move this into a separate method at least, and couldn't we call this from the event handler or something?